From: =?utf-8?b?Ik1hcmNvIFRyZXZpc2FuIChUcmV2acOxbyki?= <mail@3v1n0.net>
Date: Mon, 28 Mar 2022 15:00:22 +0200
Subject: js: Support legacy GNOME Control Center (now Settings) dbus name

Without this we won't be able to launch settings panels from shell or to
handle DBus operations from control center.

This can be safely removed post-42 (or when we will switch to
gnome-control-center 42).
---
 js/ui/shellDBus.js      |  1 +
 js/ui/status/network.js | 17 +++++++++++++++++
 2 files changed, 18 insertions(+)

diff --git a/js/ui/shellDBus.js b/js/ui/shellDBus.js
index edc16eb..41d0fc9 100644
--- a/js/ui/shellDBus.js
+++ b/js/ui/shellDBus.js
@@ -22,6 +22,7 @@ var GnomeShell = class {
         this._dbusImpl.export(Gio.DBus.session, '/org/gnome/Shell');
 
         this._senderChecker = new DBusSenderChecker([
+            'org.gnome.ControlCenter',
             'org.gnome.Settings',
             'org.gnome.SettingsDaemon.MediaKeys',
         ]);
diff --git a/js/ui/status/network.js b/js/ui/status/network.js
index 05f27fc..7e75a01 100644
--- a/js/ui/status/network.js
+++ b/js/ui/status/network.js
@@ -90,6 +90,23 @@ function launchSettingsPanel(panel, ...args) {
         'desktop-startup-id': new GLib.Variant('s',
             `_TIME${global.get_current_time()}`),
     };
+    try {
+        Gio.DBus.session.call(
+            'org.gnome.ControlCenter',
+            '/org/gnome/ControlCenter',
+            'org.freedesktop.Application',
+            'ActivateAction',
+            new GLib.Variant('(sava{sv})',
+                ['launch-panel', [param], platformData]),
+            null,
+            Gio.DBusCallFlags.NONE,
+            -1,
+            null);
+        return;
+    } catch (e) {
+        log(`Failed to launch Settings panel: ${e.message}, trying new name...`);
+    }
+
     try {
         Gio.DBus.session.call(
             'org.gnome.Settings',
